src/Microsoft.AspNetCore.Blazor.Browser/Interop/JavaScriptInvoke.cs

```diff
@@ -0,0 +1,253 @@
+                                    options.Async.FunctionName,
+                                    options.Async.ResolveId,
+                                    returnValue);
+                            }
+                        }
+                        else
+                        {
+                            RegisteredFunction.Invoke<bool>(
+                                options.Async.FunctionName,
+                                options.Async.RejectId);
+                        }
+
+                    }
+                    catch (Exception e)
+                    {
+                        Console.WriteLine(e.ToString());
+                        throw;
+                    }
+                    finally
+                    {
+                        Console.WriteLine("Finally!");
+                    }
+                });
+            }
+
+            return JsonUtil.Serialize(result);
+        }
+
+        internal static MethodInfo ResolveMethod(MethodOptions options)
+        {
+            throw new NotImplementedException();
+        }
+
+        private static Func<string, object> GetOrCreateInvoker(MethodOptions options, Func<string, object[]> argumentDeserializer)
+        {
+            var method = options.GetMethodOrThrow();
+            return (string args) => method.Invoke(null, argumentDeserializer(args));
+        }
+
+        private static Func<string, object[]> GetOrCreateArgumentDeserializer(MethodOptions options)
+        {
+            var info = options.GetMethodOrThrow();
+            var argsClass = ArgumentList.GetArgumentClass(info.GetParameters().Select(p => p.ParameterType).ToArray());
+            var deserializeMethod = ArgumentList.GetDeserializer(argsClass);
+            var toParameterListMethod = argsClass.GetMethods(BindingFlags.Instance | BindingFlags.Public)
+                .Where(m => string.Equals(nameof(ArgumentList.ToParameterList), m.Name))
+                .Single();
+
+            return Deserialize;
+
+            object[] Deserialize(string arguments)
+            {
+                var argsInstance = deserializeMethod(arguments);
+                return argsInstance.ToParameterList();
+            }
+        }
+
+        /// <summary>
+        /// TODO
+        /// </summary>
+        /// <param name="identifier"></param>
+        /// <param name="args"></param>
+        /// <returns></returns>
+        public static Task InvokeJavaScriptVoidFunctionAsync(string identifier, params object[] args)
+        {
+            var tcs = new TaskCompletionSource<int>();
+            var argsJson = args.Select(JsonUtil.Serialize);
+            var successId = Guid.NewGuid().ToString();
+            var failureId = Guid.NewGuid().ToString();
+            var asyncProtocol = JsonUtil.Serialize(new
+            {
+                Success = successId,
+                Failure = failureId,
+                Function = new MethodOptions
```

Comment:
As suggested above, let's consider not passing any `Function` value at all, since we control the protocol and know what the value is going to be. This will save the need to construct the `MethodOptions` and all the `TypeInstance` values.